src/dragonball/src/vmm.rs

```diff
@@ -0,0 +1,215 @@
+// found in the THIRD-PARTY file.
+
+use std::os::unix::io::RawFd;
+use std::sync::{Arc, Mutex, RwLock};
+
+use dbs_utils::epoll_manager::EpollManager;
+use log::{error, info, warn};
+use seccompiler::BpfProgram;
+use vmm_sys_util::eventfd::EventFd;
+
+use crate::api::v1::{InstanceInfo, VmmService};
+use crate::error::{EpollError, Result};
+use crate::event_manager::{EventContext, EventManager};
+use crate::vm::Vm;
+use crate::{EXIT_CODE_GENERIC_ERROR, EXIT_CODE_OK};
+
+/// Global coordinator to manage API servers, virtual machines, upgrade etc.
+///
+/// Originally firecracker assumes an VMM only manages an VM, and doesn't distinguish VMM and VM.
+/// Thus caused a mixed and confusion design. Now we have explicit build the object model as:
+///  |---Vmm API Server--<-1:1-> HTTP API Server
+///  |        |----------<-1:1-> Shimv2/CRI API Server
+///  |
+/// Vmm <-1:N-> Vm <-1:1-> Address Space Manager <-1:N-> GuestMemory
+///  ^           ^---1:1-> Device Manager <-1:N-> Device
+///  |           ^---1:1-> Resource Manager
+///  |           ^---1:N-> Vcpu
+///  |---<-1:N-> Event Manager
+pub struct Vmm {
+    pub(crate) event_ctx: EventContext,
+    epoll_manager: EpollManager,
+
+    // Will change to a HashMap when enabling 1 VMM with multiple VMs.
+    vm: Vm,
+
+    vcpu_seccomp_filter: BpfProgram,
+    vmm_seccomp_filter: BpfProgram,
+}
+
+impl Vmm {
+    /// Create a Virtual Machine Monitor instance.
+    pub fn new(
+        api_shared_info: Arc<RwLock<InstanceInfo>>,
+        api_event_fd: EventFd,
+        vmm_seccomp_filter: BpfProgram,
+        vcpu_seccomp_filter: BpfProgram,
+        kvm_fd: Option<RawFd>,
+    ) -> Result<Self> {
+        let epoll_manager = EpollManager::default();
+        Self::new_with_epoll_manager(
+            api_shared_info,
+            api_event_fd,
+            epoll_manager,
+            vmm_seccomp_filter,
+            vcpu_seccomp_filter,
+            kvm_fd,
+        )
+    }
+
+    /// Create a Virtual Machine Monitor instance with a epoll_manager.
+    pub fn new_with_epoll_manager(
+        api_shared_info: Arc<RwLock<InstanceInfo>>,
+        api_event_fd: EventFd,
+        epoll_manager: EpollManager,
+        vmm_seccomp_filter: BpfProgram,
+        vcpu_seccomp_filter: BpfProgram,
+        kvm_fd: Option<RawFd>,
+    ) -> Result<Self> {
+        let vm = Vm::new(kvm_fd, api_shared_info, epoll_manager.clone())?;
+        let event_ctx = EventContext::new(api_event_fd)?;
+
+        Ok(Vmm {
+            event_ctx,
+            epoll_manager,
+            vm,
+            vcpu_seccomp_filter,
+            vmm_seccomp_filter,
+        })
+    }
+
+    /// Get a reference to a virtual machine managed by the VMM.
+    pub fn get_vm_by_id(&self, _id: &str) -> Option<&Vm> {
+        Some(&self.vm)
+    }
+
+    /// Get a mutable reference to a virtual machine managed by the VMM.
+    pub fn get_vm_by_id_mut(&mut self, _id: &str) -> Option<&mut Vm> {
```

Comment:
Why are these methods used? There are lots of calls to and they all specify an empty id (`get_vm_by_id_mut("")`), so why is the id param needed? The current calls look odd so why not just call these methods `get_vm()` and `get_vm_mut()`?

Reply:
Changed.